### PR TITLE
Add point transformer attention

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+    conditioning modulates both the attention sublayer and the
+    feed-forward sublayer.
+
+    When the input has at most one point, the attention sublayer is
+    skipped and only the FFN is applied (still receiving ``cond`` if
+    provided).
+
+    Parameters
+    ----------
+    dim : int
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py:380]: This has been fixed.

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+        attn_in = (q.unsqueeze(1) - k + rel_bias).reshape(
+            int(h.shape[0]), idx.shape[1], self.dim
+        )
+        logits = self.attn_proj(attn_in).transpose(1, 2)
+        attn = _apply_neighbor_mask(logits, neighbor_mask)
+        value = (v + rel_bias).permute(0, 2, 1, 3)
+        out = (attn.unsqueeze(-1) * value).sum(dim=2).reshape(int(h.shape[0]), self.dim)
+        out = self.dropout(self.out_proj(out))
+        if gate is not None:
+            out = out * (gate if self.adaln_zero else (1.0 + gate))
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py]: Point transformer attention is extensively used in many of the modern transformer architectures for physics-ML. This is not an experimental layer/feature and can be directly added here.

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+    - ``query_features``: ``(Nq, dim)``; ``query_coords``: ``(Nq, D)``.
+    - ``context_features``: ``(Nc, dim)``; ``context_coords``: ``(Nc, D)``.
+    - ``cond`` / ``context_cond`` (if used): ``(D_cond,)`` or
+      ``(N, D_cond)``.
+    - ``query_batch_ids`` / ``context_batch_ids`` (optional): ``(Nq,)`` /
+      ``(Nc,)`` ``int64``; when both are provided, neighbors from
+      different batches are masked out of attention.
+    - Output: ``(Nq, dim)``.
+
+    Raises
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py]: This has been added.

File: test/nn/module/test_point_transformer_attention.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+import torch.nn as nn
+
+from physicsnemo.nn import (
+    LocalPointTransformerBlock,
+    LocalTokenCrossAttentionBlock,
+    ResidualMLP,
+)
+from physicsnemo.nn.module.point_transformer_attention import _dilated_knn
+
+
```
[the remark this answers is not in the document]
Comment (mnabian) [reply test/nn/module/test_point_transformer_attention.py:48]: Fixed.

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+        if gate is not None:
+            out = out * (gate if self.adaln_zero else (1.0 + gate))
+        out = residual + out
+        return self.ffn(out, cond=cond)
+
+
+class LocalTokenCrossAttentionBlock(nn.Module):
+    r"""Local cross-attention from query tokens to a per-query k-NN of context.
+
+    Each query attends to its ``neighbor_k`` nearest context tokens (by
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py:516]: Fixed.

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+            else _make_conditioning_mlp(int(conditioning_dim), 3 * self.dim)
+        )
+        self.q_proj = nn.Linear(self.dim, self.dim)
+        self.k_proj = nn.Linear(self.dim, self.dim)
+        self.v_proj = nn.Linear(self.dim, self.dim)
+        self.pos_proj = nn.Sequential(
+            nn.Linear(3, self.dim),
+            nn.GELU(),
+            nn.Linear(self.dim, self.dim),
+        )
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py]: Fixed.

File: physicsnemo/nn/module/point_transformer_attention.py [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+
+
+class LocalPointTransformerBlock(nn.Module):
+    r"""Local self-attention block over a per-point k-NN graph.
+
+    For each point, attends to its ``neighbor_k`` nearest neighbors with a
+    learned relative-position bias and per-head (grouped) vector-attention
+    scores -- the score for each neighbour is produced by an MLP applied to
+    ``query - key + positional_bias`` (Point Transformer), not a dot
+    product. Followed by a :class:`ResidualMLP`. Optional AdaLN/AdaLN-Zero
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/point_transformer_attention.py]: Fixed.

**Review by loliverhennigh — APPROVED** [review]
LGTM
